DataFormats/HGCDigi/interface/PHGCSimAccumulator.h

```diff
@@ -0,0 +1,156 @@
+#ifndef DataFormats_HGCDigi_PHGCSimAccumulator_h
+#define DataFormats_HGCDigi_PHGCSimAccumulator_h
+
+#include "DataFormats/DetId/interface/DetId.h"
+
+#include <vector>
+#include <cassert>
+
+class PHGCSimAccumulator {
+public:
+  // These two structs are public only because of dictionary generation
+  class DetIdSize {
+  public:
+    // Use top 27 bits to hold the details of the DetId
+    constexpr static unsigned detIdOffset = 5;
+    constexpr static unsigned detIdMask = 0x7ffffff;
+    // Use the last 5 bits to index 2x15 elements
+    constexpr static unsigned sizeOffset = 27;
+    constexpr static unsigned sizeMask = 0x1f;
+    // With this arrangement increasing the size component is faster
+
+    DetIdSize(): detIdSize_(0) {}
+    DetIdSize(unsigned int detId): detIdSize_(detId << detIdOffset) {}
+
+    void increaseSize() {
+      assert(size()+1 < 1<<detIdOffset);
+      ++detIdSize_;
+    }
+
+    unsigned int detIdDetails() const { return detIdSize_ >> detIdOffset; }
+    unsigned int size() const { return detIdSize_ & sizeMask; }
+
+  private:
+    unsigned int detIdSize_;
+  };
+  class Data {
+  public:
+    constexpr static unsigned energyOffset = 15;
+    constexpr static unsigned energyMask = 0x1;
+    constexpr static unsigned sampleOffset = 11;
+    constexpr static unsigned sampleMask = 0xf;
+    constexpr static unsigned dataOffset = 0;
+    constexpr static unsigned dataMask = 0x7ff;
+
+    Data(): data_(0) {}
+    Data(unsigned short ei, unsigned short si, unsigned short d):
+      data_((ei << energyOffset) | (si << sampleOffset) | d)
+    {}
+
+    unsigned int energyIndex() const { return data_ >> energyOffset; }
+    unsigned int sampleIndex() const { return (data_ >> sampleOffset) & sampleMask; }
+    unsigned int data() const { return data_ & dataMask; }
+
+  private:
+    unsigned short data_;
+  };
+
+  PHGCSimAccumulator() = default;
+  PHGCSimAccumulator(unsigned int detId): detSubdetId_(detId >> DetId::kSubdetOffset) {}
+  ~PHGCSimAccumulator() = default;
+
+  void reserve(size_t size) {
+    detIdSize_.reserve(size);
+    data_.reserve(size);
+  }
+
+  void shrink_to_fit() {
+    detIdSize_.shrink_to_fit();
+    data_.shrink_to_fit();
+  }
+
+  void emplace_back(unsigned int detId, unsigned short energyIndex, unsigned short sampleIndex, unsigned short data) {
```

Comment:
it might be confusing to have a function named `emplace_back()` that doesn't act like STL `emplace_back()` (i.e. fixed rather than variadic arguments)

Reply:
On the other hand the container is already specialized (e.g. `std::vector<int>::emplace_back()` accepts only arguments convertible(?) to `int`). Of course I can change the name if there are better suggestions.

Reply:
Maybe it's okay since other member functions of the class also do operations on the internal vectors. But this one also calls `detIdSize_.back().increaseSize();`, so could call it `append` or something.

Reply:
The `increaseSize()` only increases an internal index counter. If the class would be implemented internally as `vector<pair<unsigned, vector<...> > >` (which data sort-of resembles, the current implementation then linearizes the double-vector), would you still prefer to call it something else than `emplace_back()`?

Reply:
Eh, let's just leave it.